src/librbd/io/ImageRequestWQ.cc

```diff
@@ -605,15 +674,12 @@ void *ImageRequestWQ<I>::_void_dequeue() {
     return nullptr;
   }
 
-  if (!peek_item->was_throttled() &&
-      iops_throttle->get<
-        ImageRequestWQ<I>, ImageDispatchSpec<I>,
-        &ImageRequestWQ<I>::handle_iops_throttle_ready>(1, this, peek_item)) {
+  if (needs_throttle(peek_item)) {
     ldout(cct, 15) << "throttling IO " << peek_item << dendl;
 
-    // dequeue the throttled item and block future IO
+    // dequeue the throttled item
     ThreadPool::PointerWQ<ImageDispatchSpec<I> >::_void_dequeue();
-    ++m_io_blockers;
```

Comment:
Need to come up w/ a plan where flushes cannot be re-ordered ahead of writes. Right now, if a write is throttled, a flush in the queue can proceed and flush "nothing" and complete before the write proceeds.

Reply:
Wow, good point! I missed the flush logic.  thanx a lot @dillaman

Reply:
Hi, @dillaman , I am thinking to make flush as an IO barrier to make sure flush is waiting for writes. But I have another question, should we guarantee the order of read and write to the same location?

Reply:
@dillaman After a second thought, maybe we don't need a special logic for flush. I checked qemu and kernel about the IO throttle. 
Qemu: https://github.com/qemu/qemu/blob/master/block/throttle.c#L104
```
static int coroutine_fn throttle_co_preadv(BlockDriverState *bs,
                                           uint64_t offset, uint64_t bytes,
                                           QEMUIOVector *qiov, int flags)
{

    ThrottleGroupMember *tgm = bs->opaque;
    throttle_group_co_io_limits_intercept(tgm, bytes, false);

    return bdrv_co_preadv(bs->file, offset, bytes, qiov, flags);
}

static int coroutine_fn throttle_co_pwritev(BlockDriverState *bs,
                                            uint64_t offset, uint64_t bytes,
                                            QEMUIOVector *qiov, int flags)
{
    ThrottleGroupMember *tgm = bs->opaque;
    throttle_group_co_io_limits_intercept(tgm, bytes, true);

    return bdrv_co_pwritev(bs->file, offset, bytes, qiov, flags);
}
... ...

static int throttle_co_flush(BlockDriverState *bs)
{
    return bdrv_co_flush(bs->file->bs);
}
```
We can see that read and write will be throttled by ```throttle_group_co_io_limits_intercept(tgm, bytes, true);```. But in flush, it did nothing check and just call the real flush ```bdrv_co_flush(bs->file->bs);```

kernel: https://github.com/torvalds/linux/blob/master/block/blk-core.c#L2249
```
static noinline_for_stack bool
generic_make_request_checks(struct bio *bio)
{
	struct request_queue *q;
	int nr_sectors = bio_sectors(bio);
	blk_status_t status = BLK_STS_IOERR;
	char b[BDEVNAME_SIZE];
... ...
	if (!blkcg_bio_issue_check(q, bio))
		return false;
...
```
then https://github.com/torvalds/linux/blob/master/include/linux/blk-cgroup.h#L683
```
static inline bool blkcg_bio_issue_check(struct request_queue *q,
					 struct bio *bio)
{
	struct blkcg *blkcg;
	struct blkcg_gq *blkg;
	bool throtl = false;

	rcu_read_lock();
	blkcg = bio_blkcg(bio);

... ...

	throtl = blk_throtl_bio(q, blkg, bio);

```
There are no logic to make flush wait the writes complete.

I mean, maybe the flush means make sure the all ```returned``` write operations flushed to disk rather than all ```queued``` write operations.  Example:
```
Thread A                                Thread B

write A and returned.
flush                                   Writing B

And we got the "write B" request first in workqueue.

In this case, the flush request should only guarantee that A
will be flushed to disk. About B, it maybe flushed if it completed
before flush start, or not if it didn't put his data into disk cache,
that's not guaranteed.

But yes, we need a locking in this case to avoid
this kind of race in this example, I just put it as
an example here.
``` 
I did not find an official Semantics for flush operation. Just learn this above from reading the code of qemu and kernel. That's your opinion? thanx

Reply:
@yangdongsheng Sorry for the delay -- I've been side-tracked. 

> I did not find an official Semantics for flush operation. Just learn this above from reading the code of qemu and kernel. That's your opinion? thanx

I concur that this is the correct semantics so you are correct that you don't need to worry about throttling the flush. librbd internally just tracks the flush against all previous IOs as a way to handle IO guarding (e.g. blocking writes or completing all IOs when preparing to close the image).

Reply:
Yup, seems we need to override the drain() in ImageRequestWQ to complete all IOs when prepareing to close. :) thanx @dillaman